### PR TITLE
feat(stepfunctions):  add intrinsic functions

### LINKED ISSUE #22068: (aws-stepfunctions): Add missing intrinsic function `States.JsonMerge`
### Describe the feature

Support for the intrinsic function `States.JsonMerge` .
AWS Documentation: https://docs.amazonaws.cn/en_us/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html#asl-intrsc-func-json-manipulate

### Use Case

Obvious

### Proposed Solution

Add `JsonPath.jsonMerge()` here:
https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-stepfunctions/lib/fields.ts

### Other Information

We could use `States.JsonMerge` like below, but syntax check fails when running `cdk deploy` .
We have to change some code to make it passed.

```
TaskInput.fromJsonPathAt('States.JsonMerge($.json1, $.json2, false)')
```

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.41.0

### Environment details (OS name and version, etc.)

macOS Big Sur 11.5

### LINKED ISSUE #22629: step_functions: Support new (8/31/2022) intrinsic functions
### Describe the feature

The Amazon States Language used by Step Functions recently added a number of very useful intrinsic functions that aren't currently supported by the CDK: https://states-language.net/#2022-08-31

- States.ArrayPartition
- States.ArrayContains
- States.ArrayRange
- States.ArrayGetItem
- States.ArrayLength
- States.ArrayUnique
- States.Base64Encode
- States.Base64Decode
- States.Hash
- States.JsonMerge
- States.MathRandom
- States.MathAdd
- States.StringSplit
- States.UUID

When we try to use these functions, the CDK (Python variant) returns the following error indicating that the only supported functions are those present before the update:
RuntimeError: Error: JSON path values must be exactly '$', '$$', start with '$.', start with '$$.', start with '$[', or start with an intrinsic function: States.Format, States.StringToJson, States.JsonToString, or States.Array. Received: States.MathAdd($.successful_count, 1)

### Use Case

These updates to the States Language are a big step forward in improving the utility of Step Functions, particularly when integrating with a range of AWS services, so it would be incredibly valuable to have them available for inclusion in our workflows.

### Proposed Solution

Adding these functions to JsonPath would be useful, although a simple solution would be to add them to the allowed values shown in the error message above so that we can simply use JsonPath.string_at("States.MathAdd($.successful_count, 1)").

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.47.0

### Environment details (OS name and version, etc.)

Windows 11 and Amazon Linux

### PR DESCRIPTION
**Pull request by gkkachi**
Resolves #22068 and resolves #22629

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts [hunk 1]
```diff
@@ -312,17 +312,22 @@ function pathFromToken(token: IResolvable | undefined) {
...
+ * If the value is a string -- single-quote it.
+ * Otherwise, throw errors.
  *
  * Call this function whenever you're building compound JSONPath expressions, in
  * order to avoid having tokens-in-tokens-in-tokens which become very hard to parse.
  */
-export function renderInExpression(x: string) {
-  const path = jsonPathString(x);
-  return path ?? singleQuotestring(x);
+export function renderInExpression(x: any) {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts:325]: Rather than throwing an error here for other types and accepting any, x should be typed to the specific valid types:

```suggestion
export function renderInExpression(x: string | number) {
```

File: packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import * as cdk from '@aws-cdk/core';
+import * as integ from '@aws-cdk/integ-tests';
+import { JsonPath, Pass, StateMachine } from '../lib';
+
+/*
+ * Stack verification steps:
+ *
+ * -- aws stepfunctions describe-state-machine --state-machine-arn <stack-output> has a status of `ACTIVE`
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts:8]: This should use an assertion in code to verify instead of being added here in text.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Thank you for your contribution! Please see my inline comments.

File: packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts (hunk 1 above)
Comment (gkkachi) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts:325]: `renderInExpression` can receive `string[]` here, so it should at least accept `string[]`. (Note that `listAt` returns `string[]`. )
We could restrict the input type by creating another function like `renderListInExpression`, but I think it is not necessary because the function is a "private" function.

https://github.com/gkkachi/aws-cdk/blob/ab29124d26cf8a21a1399b0193122c7b8419d00d/packages/%40aws-cdk/aws-stepfunctions/lib/fields.ts#L113-L115
https://github.com/gkkachi/aws-cdk/blob/ab29124d26cf8a21a1399b0193122c7b8419d00d/packages/%40aws-cdk/aws-stepfunctions/test/fields.test.ts#L241

File: packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts (hunk 2 above)
Comment (gkkachi) [reply packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts]: I created this integration test based on this.
I think it is not necessary to execute `describe-state-machine` explicitly to confirm that the state machine is deployed successfully because successful deployment of cfn stack means that state machine deployments are also succeeded, so I deleted the comment by https://github.com/aws/aws-cdk/pull/22431/commits/926c938b6df88fea14409133abfd714bfc1f9fc0.

https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-stepfunctions/test/integ.custom-state.ts

File: packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts (hunk 1 above)
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts:325]: If it can accept string[], the logic isn't quite right, then. You throw an error if it's not string or number.

File: packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts (hunk 2 above)
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts]: Please add the assertion anyway. We didn't have that library when the older test was written but we do have that capability now.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Comments inline.

File: packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts (hunk 1 above)
Comment (gkkachi) [reply packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts:325]: I wrote tests for `renderInExpression` . https://github.com/aws/aws-cdk/pull/22431/commits/d32fbf397ef27c2a8fda867a30c37366ef43a93d

File: packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts (hunk 2 above)
Comment (gkkachi) [reply packages/@aws-cdk/aws-stepfunctions/test/integ.intrinsics.ts]: Ok, I added it. https://github.com/aws/aws-cdk/pull/22431/commits/d2ff6e1c2053915433f8a1aa5aed3219b766474d

**Review by TheRealAmazonKendra — APPROVED** [review]

File: packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts [hunk 3]
```diff
@@ -312,17 +312,22 @@ function pathFromToken(token: IResolvable | undefined) {
...
+  const path = jsonPathFromAny(x);
+  if (path) return path;
+  if (typeof x === 'number') return x.toString(10);
+  if (typeof x === 'string') return singleQuotestring(x);
+  throw new Error('Unxexpected value.');
```
Comment (morrijm4) [inline packages/@aws-cdk/aws-stepfunctions/lib/private/json-path.ts:330]: Extra x in unexpected
